### PR TITLE
Import github.com/mattmoor/k8schain as pkg/authn/k8schain.

### PR DESCRIPTION
**Pull request by mattmoor**
I had initially hosted this as a separate repo because this bring in so many vendor dependencies, but I believe that keeping it as a separate package will enable `dep` to prune it (and its transitive deps) reducing the downstream vendor footprint in essentially the same way.

### CODE REVIEW

**Comment by dekkagaijin** [comment]
This might complicate things when importing to googl3, though it seems like copybara would make it less of a headache (don't pull this package dir, rewrite package imports).

File: pkg/authn/k8schain/tests/serviceaccount/main.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+func main() {
+	ref, err := name.NewTag("gcr.io/build-crd-testing/secret-sauce:latest", name.WeakValidation)
+	if err != nil {
+		log.Fatalf("NewTag() = %v", err)
+	}
+
+	kc, err := k8schain.NewInCluster(k8schain.Options{
+		Namespace:          "serviceaccount-namespace",
+		ServiceAccountName: "serviceaccount",
+		// ImagePullSecrets: []string{
```
Comment (jonjohnsonjr) [inline pkg/authn/k8schain/tests/serviceaccount/main.go]: Intentionally left?

Comment (mattmoor) [reply pkg/authn/k8schain/tests/serviceaccount/main.go]: Ah, I think this simply indicates the name of the pull secret attached to the service account.  I added a clarifying comment, but it is intentionally commented out.

**Comment by mattmoor** [comment]
@dekkagaijin We already filter this when we import it.

File: pkg/authn/k8schain/k8schain.go [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+	}
+
+	client, err := kubernetes.NewForConfig(clusterConfig)
+	if err != nil {
+		return nil, err
+	}
+	return New(client, opt)
+}
+
+// NewNoClient returns a new authn.Keychain that supports the portion of the K8s keychain
```
Comment (jonjohnsonjr) [inline pkg/authn/k8schain/k8schain.go]: nit: portions

File: pkg/authn/k8schain/k8schain.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+// that don't read ImagePullSecrets.  This limits it to roughly the Node-identity-based
+// authentication schemes in Kubernetes pkg/credentialprovider.  This version of the
+// k8schain drops the requirement that we run as a K8s serviceaccount with access to all
+// of the on-cluster secrets.  This drop in fidelity also diminishes its value as a standin
```
Comment (jonjohnsonjr) [inline pkg/authn/k8schain/k8schain.go]: nit: stand-in

File: pkg/authn/k8schain/k8schain.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+type keychain struct {
+	keyring credentialprovider.DockerKeyring
+}
+
+// Resolve implements authn.Keychain
+func (kc *keychain) Resolve(reg name.Registry) (authn.Authenticator, error) {
+	// TODO(mattmoor): Lookup expects an image reference and we only have a registry,
+	// find something better than this.
+	creds, found := kc.keyring.Lookup(reg.String() + "/foo/bar")
```
Comment (jonjohnsonjr) [inline pkg/authn/k8schain/k8schain.go]: lol

**Pull request merged**